crates/ra_hir_def/src/visibility.rs

```diff
@@ -89,6 +89,44 @@ impl Visibility {
             ast::VisibilityKind::Pub => Visibility::Public,
         }
     }
+
+    pub fn resolve(
+        &self,
+        db: &impl DefDatabase,
+        resolver: &crate::resolver::Resolver,
+    ) -> ResolvedVisibility {
+        // we fall back to public visibility (i.e. fail open) if the path can't be resolved
+        resolver.resolve_visibility(db, self).unwrap_or(ResolvedVisibility::Public)
+    }
+}
+
+/// Visibility of an item, with the path resolved.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum ResolvedVisibility {
```

Comment:
Not sure if it makes sense to distinguish between resolved and unresolved visibility. I guess producing a resolved one directly might be simpler? Especially if we special case `priv`, `pub` and `pub(crate)`, which don't need to resolve paths at all.

Reply:
I think the distinction is useful during name resolution, mirroring the raw items -> defs distinction :)

Btw, I would say that `pub(crate)` and `priv` need to resolve paths in the sense of determining the crate/the current module -- if we don't put that information into the resolved visibility, we'll need it later when checking the visibility, and then every caller would have to take care to pass the correct definition module.

Reply:
It might make sense to rename `Visibility` -> `RawVisibility` and `ResolvedVisibility` -> `Visibility` though :slightly_smiling_face:

Reply:
> need to resolve paths in the sense of determining the crate/the current module

Yes, we should store `ModuleId`s. My point is that resolving `pub(crate)` and `priv` is easy, in that you don't actually need to resolve an `hir::Path`, you need only the module structure.

Reply:
(I just did the rename using the LSP functionality, and the only reference it missed was the one in the `salsa::invoke` annotation -- maybe I shouldn't be impressed, but I am :slightly_smiling_face: )

Reply:
And note that the one in `salsa::invoke` can't handled correctly even in theory. It's a sad truth that a language with macros can't have precise rename because `macro_rules! fork_you { () => { $crate::unrenamable::path() } }` :(